From 77235e3c6ab782c3230e80b6582649c62c80a7d4 Mon Sep 17 00:00:00 2001
From: Jason <jasonanderson@uchicago.edu>
Date: Thu, 11 Apr 2019 15:41:17 -0500
Subject: [PATCH 2/2] Detect region when using versionless id endpoints

When using versionless Keystone endpoints in the service catalog (which
is the recommended way as it allows the client to do version selection),
region detection was broken. The reason is that a v3 suffix was
automatically added to the identity service URL before doing the region
detection. Region detection works by querying the service catalog for
which endpoint matches the Keystone auth URL used for the request. Since
the endpoint URL does not have a version suffix, the search will never
succeed, so the region defaults to None, and then later the first
endpoint is picked as a fallback behavior.

This fixes the check to also look for the original, unfixed URL when
trying to detect the region.
---
 openstack_auth/backend.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/openstack_auth/backend.py b/openstack_auth/backend.py
index 5be15e810..60f7c3044 100644
--- a/openstack_auth/backend.py
+++ b/openstack_auth/backend.py
@@ -93,6 +93,7 @@ class KeystoneBackend(object):
         if not auth_url:
             auth_url = settings.OPENSTACK_KEYSTONE_URL
 
+        orig_auth_url = auth_url
         auth_url, url_fixed = utils.fix_auth_url_version_prefix(auth_url)
         if url_fixed:
             LOG.warning("The OPENSTACK_KEYSTONE_URL setting points to a v2.0 "
@@ -167,7 +168,9 @@ class KeystoneBackend(object):
         id_endpoints = scoped_auth_ref.service_catalog.\
             get_endpoints(service_type='identity')
         for id_endpoint in [cat for cat in id_endpoints['identity']]:
-            if auth_url in id_endpoint.values():
+            # Simple way to support v2.0 and v3 Keystone service catalogs
+            search_list = id_endpoint.values()
+            if auth_url in search_list or orig_auth_url in search_list:
                 region_name = id_endpoint['region']
                 break
 
-- 
2.20.1

